### ISSUE TITLE
Multiline strings don't line up

### ISSUE THREAD

**Issue by willnwhite** [avh4/elm-format#219]
![screen shot 2016-08-03 at 15 28 47](https://cloud.githubusercontent.com/assets/8081877/17369192/35d7f7de-598f-11e6-8ba3-7a864cf30589.png)

**Comment by avh4** [comment]
Do you have a suggestion for how this should work?

In your example, the content of the string is `<newline><space>`.  elm-format does not change the contents of the string, (and rightly so, I believe).  Is there a way we can format this better without changing the content of the string?

**Comment by willnwhite** [comment]
This is what I typed (no space, just newline): 

![screen shot 2016-08-19 at 09 31 01](https://cloud.githubusercontent.com/assets/8081877/17803678/c97a14ec-65ef-11e6-9760-54ae90c861d1.png)

I think elm-format should move them both over, rather than just the top one.

**Comment by Janiczek** [comment]
What @avh4 said is that moving them both over would change the contents of the string.
```elm
string1 =
  """
  """

string2 =
    """
    """
```

These two aren't the same.
```
string1 == "\n  "
string2 == "\n    "
```

**Comment by willnwhite** [comment]
Okay, I didn't know how multiline strings work. How does moving neither over sit?

**Comment by willnwhite** [comment]
Here's a real example. I've pasted a multiline string in:

![screen shot 2016-08-19 at 11 25 07](https://cloud.githubusercontent.com/assets/8081877/17807020/047b839a-6600-11e6-862a-3e408a778981.png)

Then I've run elm-format (on save):

![screen shot 2016-08-19 at 11 25 21](https://cloud.githubusercontent.com/assets/8081877/17807022/0a9e3a92-6600-11e6-9a5c-787e17a29731.png)

It aligns the first line of the string but not the rest.

**Comment by Janiczek** [comment]
**tl;dr: `elm-format` shouldn't change the values**, that's why it'll rather leave the `"""`s unaligned than mess with them

----

I think there is no safe way to align both the same way everytime.

In your case it could be OK to deindent the spaces and not lose information (and the program would look nice):

```diff
query =
-    """
-        SELECT ?cause ... {
-          wdt:P1542 ...
-        }
-        """
+    """
+    SELECT ?cause ... {
+      wdt:P1542 ...
+    }
+    """
```

but generally there's no guarantee the user won't interpret the spaces somehow. Then `elm-format` would change what the program **does**, not only how it looks!

**Comment by willnwhite** [comment]
Okay, what I should have said is that I think it should _preserve_ the alignment.

**Comment by Janiczek** [comment]
So, either:

----
### align to self
```elm
let
    abc =
        "normal indentation"

    def =
            """
            whatever the user types, first quotes dictated by the last quotes

            + automatic
            + aligned to self
            + doesn't change string contents

            - misaligned to other definitions
            """
in
    1
```
----
### align 4 spaces
```elm
let
    abc =
        "normal indentation"

    def =
        """
            indent the first quotes 4 spaces

            + automatic
            + aligned to other definitions
            + doesn't change string contents

            - misaligned to self (OCD person - aren't we all - would change that afterwards)
            """
in
    1
```

Other than that, just a practical note, the multiline strings most likely don't look that nice. More like:

```elm
myMarkdown =
    """
# My heading

Some text with [link](http://google.com) and whatnot.
"""
```

That would make the "align to self" decision align it this way:

```elm
myMarkdown =
"""
# My heading

Some text with [link](http://google.com) and whatnot.
"""
```

which I'm not even sure is legal syntax.

**Comment by willnwhite** [comment]
I prefer **align to self**, but I still don't understand why elm-format can't do **align 4 spaces** then correct the misalignment, as I would do myself. Perhaps I don't understand how elm-format works.

**Comment by avh4** [comment]
The problem is that if elm-format does the alignment automatically, then that means that if someone wants to have a literal string that elm-format thinks is misaligned, then there's no way for the file to represent that exact string.  It would be nice if elm-format could do this alignment, but I don't see a rule that would make that possible that also wouldn't make certain other strings impossible to have in your source code.

I think the auto-alignment fixing would be a feature that could be implement in IDE plugins, however, so that the user could choose when and when not to perform the realignment.
